editor/src/app/app-state/app-state.service.ts

```diff
@@ -30,6 +31,52 @@ export class AppStateService {
     private router: Router) {
   }
 
+  showLoading() {
+    this.store.dispatch(new AppShowLoading());
+  }
+
+  hideLoading() {
+    this.store.dispatch(new AppHideLoading());
+  }
+
+  sync(urlName: string, data: any, method?: string) {
+
+    return this.store.select(state => state.app)
+      .pipe(
+        filter(appState => !!appState.user.token && appState.urls[urlName]),  // Make sure user is logged in
+        take(1),
+        map(state => {
```

Comment:
This construction of async calls is incorrect. You should use `switchMap` here and flatten the returned promise, if you use fetch.
```ts
...,
switchMap(state => {
   this.showLoading();
   return fetch(...);
}),
// then use `map` istead of `then`
map(response => {
   this.hideLoading();
   ...
}),
// then catch error
catchError(error => {
   this.hideLoading();
   ...
})
```